ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/ParseReportingUrl.java

```diff
@@ -218,23 +230,33 @@ private String getOsParam(String userAgentOs) {
   @VisibleForTesting
   List<Set<String>> loadAllowedUrls() throws IOException {
     if (singletonAllowedImpressionUrls == null || singletonAllowedClickUrls == null) {
-      singletonAllowedImpressionUrls = new HashSet<>();
-      singletonAllowedClickUrls = new HashSet<>();
+      Set<String> allowedImpressionUrls = new HashSet<>();
+      Set<String> allowedClickUrls = new HashSet<>();
```

Comment:
Having the assignment here seemed to be causing some concurrency issues in the new test where some inputs would return from this method with the empty allow list set and the message would be rejected even if the url was valid.  Looking at the `InvalidUrlException`'s in the prod errors table this look like it happened to 6 messages in the last 14 days.  This change still isn't thread-safe but it seems to work.  I'm not sure if any synchronization is needed or how to reason about that in beam

Reply:
I think this change is sufficient. It ensures that the singletons will only ever be set to null or to a complete set. If a thread sees the singleton as null, it will be guaranteed to go through the loading process again. If a thread sees a non-null value, it's guaranteed to be a complete set.

In the previous version of the code, it makes sense that a thread could end up reading a non-null but incomplete value. In GeoCityLookup, we do have `getOrCreate*` methods that are marked as synchronized which ensures we don't create those large maps multiple times in memory.